### PR TITLE
STOR-947: support disabling default StorageClass via ClusterCSIDriver

### PR DESCRIPTION
**Pull request by dobsonj**
https://issues.redhat.com/browse/STOR-947
Marking this WIP until openshift/library-go#1441 merges.
/cc @openshift/storage

### CODE REVIEW

File: pkg/operator/starter.go [hunk 1]
```diff
@@ -65,6 +67,13 @@ func RunOperator(ctx context.Context, controllerConfig *controllercmd.Controller
 		return err
 	}
 
+	// operator.openshift.io client, used for ClusterCSIDriver
+	operatorClientSet, err := opclient.NewForConfig(controllerConfig.KubeConfig)
+	if err != nil {
+		return err
+	}
+	operatorInformers := opinformers.NewSharedInformerFactory(operatorClientSet, time.Minute)
```
Comment (dobsonj) [inline pkg/operator/starter.go:75]: 1 minute resync seems kind of agressive for an object that typically doesn't change, but the alternative is that a change to StorageClassState could take a long time to take effect.

**Comment by dobsonj** [comment]
Basic functional test with the latest commit:
```
# GCP

## patch SC, should get reverted by default
$ oc patch sc standard-csi -p '{"allowVolumeExpansion":false}'
storageclass.storage.k8s.io/standard-csi patched
$ oc get sc standard-csi -o yaml | grep allowVolumeExpansion
allowVolumeExpansion: true

## Unmanaged: SC changes should persist
$ oc edit clustercsidriver pd.csi.storage.gke.io
#  storageClassState: Unmanaged
clustercsidriver.operator.openshift.io/pd.csi.storage.gke.io edited
$ oc patch sc standard-csi -p '{"allowVolumeExpansion":false}'
storageclass.storage.k8s.io/standard-csi patched
$ oc get sc standard-csi -o yaml | grep allowVolumeExpansion
allowVolumeExpansion: false

## Removed: SC should be removed
$ oc edit clustercsidriver pd.csi.storage.gke.io
#  storageClassState: Removed
clustercsidriver.operator.openshift.io/pd.csi.storage.gke.io edited
$ oc get sc
No resources found

## Managed: SC should be recreated. Changes should be reverted.
$ oc edit clustercsidriver pd.csi.storage.gke.io
#  storageClassState: Managed
clustercsidriver.operator.openshift.io/pd.csi.storage.gke.io edited
$ oc get sc
NAME                     PROVISIONER             RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
ssd-csi                  pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   2s
standard-csi (default)   pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   2s
$ oc patch sc standard-csi -p '{"allowVolumeExpansion":false}'
storageclass.storage.k8s.io/standard-csi patched
$ oc get sc standard-csi -o yaml | grep allowVolumeExpansion
allowVolumeExpansion: true
```

**Comment by dobsonj** [comment]
/retitle STOR-947: support disabling default StorageClass via ClusterCSIDriver

**Comment by ropatil010** [comment]
/label qe-approved

Payload: 4.13.0-0.ci.test-2023-01-24-042936-ci-ln-z5ttxwb-latest
Nodes version: v1.25.2+7dab57f
Working as expected.

Results:
###################test1_Removed started #################
oc get clustercsidriver pd.csi.storage.gke.io -o yaml | grep 'storageClassState'
  storageClassState: Removed

 Removed test change standard-csi allowVolumeExpansion to false
 Removed test change ssd-csi allowVolumeExpansion to false

 Removed result oc get sc
###################test2_Managed started #################
oc get clustercsidriver pd.csi.storage.gke.io -o yaml | grep 'storageClassState'
  storageClassState: Managed

oc get sc
NAME                          PROVISIONER     RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
ssd-csi                           pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   41s
standard-csi (default)   pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   41s

 change standard-csi volumeBindingMode
 change standard-csi reclaimPolicy
 change ssd-csi volumeBindingMode
 change ssd-csi reclaimPolicy

 Managed test change standard-csi allowVolumeExpansion to false
storageclass.storage.k8s.io/standard-csi patched

 Managed result oc get sc standard-csi
NAME                     PROVISIONER             RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
standard-csi (default)   pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   51s

 Managed test change ssd-csi allowVolumeExpansion to false
storageclass.storage.k8s.io/ssd-csi patched

 Managed result oc get sc ssd-csi
NAME      PROVISIONER             RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
ssd-csi   pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   54s

oc get sc
NAME                     PROVISIONER             RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
ssd-csi                  pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   55s
standard-csi (default)   pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   55s

oc delete sc ssd-csi
storageclass.storage.k8s.io "ssd-csi" deleted
storageclass.storage.k8s.io "standard-csi" deleted

oc get sc
NAME                     PROVISIONER             RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
ssd-csi                           pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   4s
standard-csi (default)   pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   2s
###################test3_Unmanaged started #################
oc get clustercsidriver pd.csi.storage.gke.io -o yaml | grep 'storageClassState'
  storageClassState: Unmanaged

oc get sc
NAME                     PROVISIONER             RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
ssd-csi                         pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   2m50s
standard-csi (default)   pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   2m48s

 change standard-csi volumeBindingMode
 change standard-csi reclaimPolicy
 change ssd-csi volumeBindingMode
 change ssd-csi reclaimPolicy

Unmanaged test change standard-csi allowVolumeExpansion to false
storageclass.storage.k8s.io/standard-csi patched

 Unmanaged result oc get sc standard-csi
NAME                     PROVISIONER             RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
standard-csi (default)   pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   false                  2m58s

Unmanaged test change ssd-csi allowVolumeExpansion to false
storageclass.storage.k8s.io/ssd-csi patched

 Unmanaged result oc get sc ssd-csi
NAME      PROVISIONER             RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
ssd-csi   pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   false                  3m3s

oc get sc
NAME                     PROVISIONER             RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
ssd-csi                           pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   false                  3m4s
standard-csi (default)   pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   false                  3m2s

oc delete sc ssd-csi
storageclass.storage.k8s.io "ssd-csi" deleted
storageclass.storage.k8s.io "standard-csi" deleted

oc get sc
###################test4a_Multiple default sc started #################
oc get clustercsidriver pd.csi.storage.gke.io -o yaml | grep 'storageClassState'
  storageClassState: Managed

oc get sc
NAME                     PROVISIONER             RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
ssd-csi                           pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   5m37s
standard-csi (default)   pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   5m35s
Creating sc
storageclass.storage.k8s.io/testsc created
Applying patch to sc testsc for changing to default
storageclass.storage.k8s.io/testsc patched

oc get sc
NAME                     PROVISIONER             RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
ssd-csi                            pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   5m40s
standard-csi (default)   pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   5m38s
testsc (default)              pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   2s
###################test4b_Removed started #################
oc get clustercsidriver pd.csi.storage.gke.io -o yaml | grep 'storageClassState'
  storageClassState: Removed

oc get sc and check ssd-csi, standard-csi should not be there, testsc should be there
NAME               PROVISIONER       RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
testsc (default)   pd.csi.storage.gke.io    Delete          WaitForFirstConsumer   true                   3m10s
###################test4c_Managed sc started #################
oc get clustercsidriver pd.csi.storage.gke.io -o yaml | grep 'storageClassState'
  storageClassState: Managed

oc get sc all should be present now ssd-csi, standard-csi, testsc
NAME               PROVISIONER             RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
ssd-csi               pd.csi.storage.gke.io    Delete          WaitForFirstConsumer   true                   8s
standard-csi       pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   8s
testsc (default)   pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   3m48s
Deleting all sc
storageclass.storage.k8s.io "testsc" deleted
storageclass.storage.k8s.io "ssd-csi" deleted
storageclass.storage.k8s.io "standard-csi" deleted

oc get sc ssd-csi and standard-csi should be back to default
NAME                     PROVISIONER             RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
ssd-csi                           pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   3s
standard-csi (default)   pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   2s
###################test5a_Managed default sc started #################
oc get clustercsidriver pd.csi.storage.gke.io -o yaml | grep 'storageClassState'
  storageClassState: Managed

oc get sc
NAME                     PROVISIONER             RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
ssd-csi                           pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   5m46s
standard-csi (default)   pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   5m45s
###################test5b_Unmanaged default sc started #################
oc get clustercsidriver pd.csi.storage.gke.io -o yaml | grep 'storageClassState'
  storageClassState: Unmanaged

oc get sc
NAME                     PROVISIONER             RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
ssd-csi                           pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   6m19s
standard-csi (default)   pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   6m18s
Creating sc
storageclass.storage.k8s.io/testsc created
Applying patch to sc testsc for changing to default
storageclass.storage.k8s.io/testsc patched
Applying patch to sc standard-csi for changing to non default
storageclass.storage.k8s.io/standard-csi patched

oc get sc
NAME               PROVISIONER             RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
ssd-csi                pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   6m24s
standard-csi       pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   6m23s
testsc (default)   pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   4s
###################test5c_Removed sc started #################
oc get clustercsidriver pd.csi.storage.gke.io -o yaml | grep 'storageClassState'
  storageClassState: Removed

oc get sc ssd-csi, standard-csi should not be there, testsc should be there
NAME               PROVISIONER       RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
testsc (default)   pd.csi.storage.gke.io   Delete          WaitForFirstConsumer   true                   45s
Deleting sc testsc
storageclass.storage.k8s.io "testsc" deleted
No sc should remain

**Comment by dobsonj** [comment]
`error: unable to read image registry.build04.ci.openshift.org/ci-op-p19rhlyi/stable@sha256:4a6ecaf7b68d463869afb275f56d4c2cbfea2bbf295d8a599bc15399ba8e3406: received unexpected HTTP status: 500 Internal Server Error`
/retest

**Comment by dobsonj** [comment]
`reason/BackOff Back-off pulling image "ghcr.io/opsgenie/kubernetes-event-exporter:v0.11` :(
/retest

**Pull request merged**